### DISCUSSION TITLE
Script tag not working

### DISCUSSION THREAD

**Discussion by bamjad** [vercel/next.js#35064] [Help]
### Verify canary release

- [ ] I verified that the issue exists in Next.js canary release

### Provide environment information

Script tag is not loading external javascripts. No matter where i place the tag. _app.js or anywhere else. 

### What browser are you using? (if relevant)

Chrome

### How are you deploying your application? (if relevant)

Dev

### Describe the Bug

Script tag is not loading external javascripts. No matter where i place the tag. _app.js or anywhere else.

1. Create a simple page
2. Import script/next
3. Use any third party js like feather icons.
4. Use google developer tools to check if script is loaded. 

### Expected Behavior

Script should be loaded

### To Reproduce

1. Create a simple page
2. Import script/next
3. Use any third party js like feather icons.
4. Use google developer tools to check if script is loaded.

**Comment by balazsorban44** [comment]
Please read the docs: https://nextjs.org/docs/api-reference/next/script#onload

**Reply by ardok** [reply]
I'm facing the same issue. This is not about `onload` callback.
```
// _document.js

import { Html, Head, Main, NextScript } from 'next/document'
import Script from 'next/script'
import React from "react";

export default function Document() {
  return (
    <Html>
      <Head />
      <body>
        <Main />
        <NextScript />
        <Script src="https://unpkg.com/ionicons@5.5.2/dist/ionicons/ionicons.esm.js" onLoad={() => console.log('hey')} />
        <Script src="https://unpkg.com/ionicons@5.5.2/dist/ionicons/ionicons.js" />
      </body>
    </Html>
  )
}
```

`hey` doesn't get printed. And the script itself is nowhere to be found on the rendered element.

----

Though it works inside `_app.js` for me. I just started with Next.js, maybe my understanding of `_document.js` is completely wrong.

**Reply by icyJoseph** [reply]
Hi,

In `_document` you cannot have event listeners, because it is evaluated on the server.

**Comment by valcosmos** [comment]
Recently, I encountered a similar problem, but I solved it with a native script.

It seems to work.

![image](https://user-images.githubusercontent.com/53572196/177237846-4f546358-84a5-44dd-affe-68a3815c25ab.png)


```js
import { Html, Head, Main, NextScript } from 'next/document'
import React from 'react'

export default function Document() {
  return (
    <Html>
      <Head></Head>
      <body>
        <Main />
        <NextScript />
        <script
          async
          type="module"
          src="https://unpkg.com/ionicons@5.5.2/dist/ionicons/ionicons.esm.js"
        />
      </body>
    </Html>
  )
}

```

**Reply by icyJoseph** [reply]
Hi, 

In Next 12.1.6, there was a bug with the Script inside the head of `_document`, could you please retry with next 12.2.0?

**Reply by valcosmos** [reply]
Hello, I am using the latest version, what is the bug u encountered, can you take a screenshot to see it?

In addition, the native script I use, not next/script.

**Reply by icyJoseph** [reply]
Hi,

This was a bug in `next/script` version 12.1.6:

- https://github.com/vercel/next.js/pull/37000
- https://github.com/vercel/next.js/issues/37225#issuecomment-1142390675
- https://github.com/vercel/next.js/pull/37084

It has been fixed in 12.2.0, AFAIK.

You shouldn't need a native script in `_document` as you are doing. A simple `next/script` should be able to take care of this. For instance:

```tsx
import "../styles/globals.css";
import type { AppProps } from "next/app";
import Script from "next/script";

function MyApp({ Component, pageProps }: AppProps) {
  return (
    <>
      <Script
        async
        src="https://unpkg.com/ionicons@5.5.2/dist/ionicons/ionicons.esm.js"
        type="module"
      />

      <Component {...pageProps} />
    </>
  );
}

export default MyApp;
```
![Screenshot 2022-07-05 at 16 20 34](https://user-images.githubusercontent.com/21013447/177349970-1e154e9d-b826-4140-ba95-32eaa40a5c1c.png)

**Reply by valcosmos** [reply]
Wow, that's awesome, I tried as you said, it does work, I didn't know that before.

Thank you so much!!😊

**Reply by tofuwu-dev** [reply]
> You shouldn't need a native script in `_document` as you are doing. A simple `next/script` should be able to take care of this. For instance:
> 
> ```tsx
> import "../styles/globals.css";
> import type { AppProps } from "next/app";
> import Script from "next/script";
> 
> function MyApp({ Component, pageProps }: AppProps) {
>   return (
>     <>
>       <Script
>         async
>         src="https://unpkg.com/ionicons@5.5.2/dist/ionicons/ionicons.esm.js"
>         type="module"
>       />
> 
>       <Component {...pageProps} />
>     </>
>   );
> }
> 
> export default MyApp;
> ```

Hi @icyJoseph 
We cant use any next script in _document file??

**Reply by icyJoseph** [reply]
> ### [beforeInteractive](https://nextjs.org/docs/basic-features/script#beforeinteractive)
> Scripts that load with the beforeInteractive strategy are injected into the initial HTML from the server and run before self-bundled JavaScript is executed. This strategy should be used for any critical scripts that need to be fetched and executed before any page becomes interactive. This strategy only works inside _document.js and is designed to load scripts that are needed by the entire site (i.e. the script will load when any page in the application has been loaded server-side).

https://nextjs.org/docs/basic-features/script#beforeinteractive

> We cant use any next script in _document file??

I didn't say that, `beforeInteractive` scripts ought to go there, `_document`. If you put a `next/script` without strategy in the `_document` file, it won't work.

If you specify `beforeInteractive` then this is loaded as part of the critical chain of resources.

If you specify `afterInteractive`, or `lazyOnload` (even though strategy is `afterInteractive` by default), then it will load, but the `onLoad` callback won't happen because it is inside `_document`.

So, why put those in `_document` if not absolutely necessary, `beforeInteractive`.

At least this is state of things at 12.2.0, let's see how this evolves together with concurrent rendering.

**Reply by tofuwu-dev** [reply]
Thanks for your good explanation @icyJoseph 
But I had some issues with Script(strategy is `afterInteractive`), google analytics doesnt work good as expected.
```bash
...
<Html lang="en">
        <Head />
        <body>
          <Main />
          <NextScript />
          <Script
            src={`https://www.googletagmanager.com/gtag/js?id=${GOOGLE_ANALYTICS_KEY}`}
            strategy="afterInteractive"
          />
          <Script id="google-analytics" strategy="afterInteractive">
            {`
                window.dataLayer = window.dataLayer || [];
                function gtag(){window.dataLayer.push(arguments);}
                gtag('js', new Date());
                gtag('config', ${GOOGLE_ANALYTICS_KEY});
              `}
          </Script>
        </body>
</Html>
...
```

**Reply by icyJoseph** [reply]
Which Next.js version did you use? With 12.1.6, that wouldn't work, and there's a little error:

```tsx
 <Script id="google-analytics" strategy="afterInteractive">
            {`
                window.dataLayer = window.dataLayer || [];
                function gtag(){window.dataLayer.push(arguments);}
                gtag('js', new Date());
                gtag('config', '${GOOGLE_ANALYTICS_KEY}');
              `}
          </Script>
```

Notice the quotes around `'${GOOGLE_ANALYTICS_KEY}'`.

Anyhow, with Next 12.2.0, this should work:

```tsx
import { Html, Head, Main, NextScript } from "next/document";
import Script from "next/script";

export default function Document() {
  return (
    <Html>
      <Head />
      <body>
        <Main />
        <NextScript />
        <Script
          src="https://cdnjs.cloudflare.com/ajax/libs/lodash.js/4.17.20/lodash.min.js"
          strategy="beforeInteractive"
        ></Script>

        <Script
          src={`https://www.googletagmanager.com/gtag/js?id=${process.env.NEXT_PUBLIC_KEY}`}
          strategy="afterInteractive"
        />
        <Script id="google-analytics" strategy="afterInteractive">
          {`
                window.dataLayer = window.dataLayer || [];
                function gtag(){window.dataLayer.push(arguments);}
                gtag('js', new Date());
                console.log("hello world")
                gtag('config', '${process.env.NEXT_PUBLIC_KEY}');
              `}
        </Script>
      </body>
    </Html>
  );
}
```

I threw in a `console.log` to convince you 😄

**Reply by icyJoseph** [reply]
@valcosmos btw in your use case, with the icons, one could also load that as `'beforeInteractive'` in `_document` to make sure they are available from the get go. It depends on the use case.

**Reply by tofuwu-dev** [reply]
> Notice the quotes around `'${GOOGLE_ANALYTICS_KEY}'`.

Do I need to add single quote?

> I threw in a console.log to convince you 😄

I could get the result of console on my end as well.. but weird... google analytics doesnt work...  :(

**Reply by icyJoseph** [reply]
Single or double should be ok, it's just that you are templating the script with `` so everything has to be correct. Without quotes it'll try to look for a window variable named after your `GOOGLE_ANALYTICS_KEY`

mmm that's really weird.

**Reply by tofuwu-dev** [reply]
Yes, btw thank you very much for your quick reply, @icyJoseph

**Reply by theWiseAman** [reply]
> Notice the quotes around `'${GOOGLE_ANALYTICS_KEY}'`.

Thanks mate this was really helpful.

**Comment by boylec** [comment]
None of the above has worked for my case which is very similar albeit I'm using slightly different inline javascript.

@latest version of next.

```jsx
import NextDocument, { Head, Html, Main, NextScript } from "next/document";
import Script from "next/script";

type Props = {};

class Document extends NextDocument<Props> {
  render() {
    return (
      <Html lang="en-US">
        <Head />
        <body>
          <Main />
          <NextScript />
          <Script id="test" strategy="afterInteractive">
            {`
              debugger;
              `}
          </Script>
          {/* Google Tag Manager */}
          <Script id="google-tag-manager" strategy="afterInteractive">
            {`
              (function(w,d,s,l,i){w[l]=w[l]||[];w[l].push({'gtm.start':
              new Date().getTime(),event:'gtm.js'});var f=d.getElementsByTagName(s)[0],
              j=d.createElement(s),dl=l!='dataLayer'?'&l='+l:'';j.async=true;j.src=
              'https://www.googletagmanager.com/gtm.js?id='+i+dl+ '&gtm_auth=${process.env.NEXT_PUBLIC_GTM_ENV_AUTH_KEY}&gtm_preview=${process.env.NEXT_PUBLIC_GTM_PREVIEW}&gtm_cookies_win=x';f.parentNode.insertBefore(j,f);
              })(window,document,'script','dataLayer','${process.env.NEXT_PUBLIC_GTM_ID}');
            `}
          </Script>
          {/* Google Tag Manager (noscript) */}
          <noscript
            dangerouslySetInnerHTML={{
              __html: `<iframe src="https://www.googletagmanager.com/ns.html?id=${process.env.NEXT_PUBLIC_GTM_ID}&gtm_auth=${process.env.NEXT_PUBLIC_GTM_ENV_AUTH_KEY}&gtm_preview=${process.env.NEXT_PUBLIC_GTM_PREVIEW}&gtm_cookies_win=x" height="0" width="0" style="display:none;visibility:hidden"></iframe>`,
            }}
          />
        </body>
      </Html>
    );
  }
}

export default Document;
```

* `debugger;` line does not fire in first <Script> section
* google-tag-manager JS in second <Script> section is not rendered
* <noscript ... /> section is rendered

Any tips?

**Reply by boylec** [reply]
# Couple notes after some testing

## TL; DR:
When I use "afterInteractive" strategy and place my component in some other component besides _document it gets evaluated. Otherwise it doesn't. "beforeInteractive" strategy seems to cause my Script element not to render at all because there is no HeadManagerContext defined at build time.

Any ideas why placing my component in _document would cause the "afterInteractive" effect not to fire in script.js at runtime (breakpoint at line 185 in image below)? In fact the Script component never runs a render cycle at all (let alone the useEffect for the "afterInteractive" strategy). This is not the case when placing the component somewhere besides _document.

Furthermore, should I not be able to use "beforeInteractive" strategy?

## Details
I used the following trivial Script definition to test with:
```
<Script id="test" strategy="afterInteractive">
  {`
    debugger;
  `}
</Script>
```

I ran my app with the element above using different strategies and placing it at different locations in the app (in/outside of _document) and got different results.

### Reference image for breakpoints (script.js in next library):
<img width="1298" alt="image" src="https://user-images.githubusercontent.com/8148174/179268767-19b68995-a4d1-4df7-87fd-6209ec108a98.png">

### Note
I use "_document.page.tsx" instead of "_document.tsx" naming scheme (*.page.tsx for all my pages)

### Tests
Conditions:
* Place element in _document
  * Use beforeInteractive strategy using next export or next dev
    * Build time (unexpected result occurs):
      * Expected result: _headManagerContext.HeadManagerContext is defined and one of lines 196, 206, or 208 would be hit.
      * Actual result: _headManagerContext.HeadManagerContext is undefined and none of the lines are hit, resulting in a null script (breakpoint at line 211 hit)
    * Run time result (unexpected result occurs):
      * Expected result: Script is rendered as part of the raw page and debugger; statement inside Script element breaks.
      * Actual result: Script is not rendered at all.
  * Use afterInteractive strategy (next dev or next export)
    * Build time (expected result occurs): Breakpoint at 184 is hit at build time. This seems expected since its "afterInteractive" strategy nothing would get evaluated at build time.
    * Run time (unexpected result occurs):
      * Expected result: Breakpoint at line 185 hits because the effect fires.
      * Actual result: Breakpoint at line 185 never hits, and in fact no render cycle happens for the Script component at all (breakpoint at line 181 is never hit either)
* Place element in some other top level component (not _app or _document). Forced to use "afterInteractive" because  "beforeInteractive" is intended for SSR/build time. Using next dev or next export.
  * Build time (expected result occurs): Same result as having the element in _document. (described above)
  * Run time (expected result occurs):  Breakpoint at line 185 hits because the effect fires. (The same expected result of placing the element in _document at runtime described above).

### Summary
Unexpecteds:
* Placing the element in _document w/ beforeInteractive strategy has unexpected results for both run time and build time.
* Placing the element in _document w/ afterInteractive strategy has **un**expected results for run time.

Can anyone speak to these unexpected results?

I.e. for bullet point 1: Is it expected that HeadManagerContext always be undefined when running next dev or next export?

For bullet point 2: Is it expected that the <Script> component itself never has a render cycle at runtime and so the effect never fires to load the script contents (whenver the <Script> is in _document)

**Reply by jasiene** [reply]
Hey @boylec, I run into the same issue as you and am using version 12.2.0

For now I've stuck all of my Script components at the top of my _app.tsx and they seem to be loading appropriately. All the scripts use the afterInteractive strategy.

Unfortunately no scripts seem to load when I stick them inside my custom _document.tsx, whether that is inside of the Head or outside, just after NextScript.

**Reply by GabrielLourenco** [reply]
Hello guys. I experienced this bug today, unfortunately. I my case, my script tags were inside a condition, as the image below:
![image](https://user-images.githubusercontent.com/17507538/179578326-6158cf46-4671-4f3f-9e95-b5e6f208c27a.png)
As you can see, it have some scripts with a mix of strategies.

After search for a intire morning, I decide remove the condition and the **react fragments**. And, for my surprise, it works as expected.

I updated my application to next v12.2.2

**Reply by Emiliano-Bucci** [reply]
@GabrielLourenco It seems that <Script /> doesn't work inside a fragment

**Reply by ZHaas** [reply]
> @GabrielLourenco It seems that <Script /> doesn't work inside a fragment

3 years later and this was finally the solution to my problem. Thanks!

**Comment by cesars-gh** [comment]
I faced the same problem with version `12.2.1`, my second script was not even rendered on the page.
What solved my issue was to set `strategy="lazyOnload"`.

```jsx
// _document.tsx
<Html>
  <Head>
    <link rel="icon" href="/favicon.ico" sizes="any" />
  </Head>
  <body>
    <Main />
    <NextScript />

    <Script id="onload-script" strategy="afterInteractive">
      {`console.log('Hello');`}
    </Script>

    <Script
      id="analytics-script"
      src="https://static.analytics.com/101375946.js"
      strategy="lazyOnload"
    />
  </body>
 </Html>
```

**Reply by mikechabotcandy** [reply]
Using 12.1.6:

After encountering and attempting to debug similar issues noted by others above, I replaced my entire existing `_document.tsx` with the suggested approach from the docs (https://nextjs.org/docs/basic-features/script#beforeinteractive). 

`lodash.min.js` script is never injected into the document

```jsx
// https://nextjs.org/docs/basic-features/script#beforeinteractive

import { Html, Head, Main, NextScript } from 'next/document'
import Script from 'next/script'

export default function Document() {
  return (
    <Html>
      <Head />
      <body>
        <Main />
        <NextScript />
        <Script
          src="https://cdnjs.cloudflare.com/ajax/libs/lodash.js/4.17.20/lodash.min.js"
          strategy="beforeInteractive"
        ></Script>
      </body>
    </Html>
  )
}
```

**Reply by GitToby** [reply]
I wasn't able to get any `beforeInteractive` scripts to load, in the end I moved all my script tags to `_app.jsx` - pre-interactivity isn't a hard requirement for my loads so it works fine. 

If `beforeInteractive` is a hard dep i cant help but if not, try moving the `<Script/>` to `_app.jsx`?

**Comment by happylolonly** [comment]
Yes, very annoying issue was, need to fix it, unclear

**Comment by spreadpando** [comment]
Also experiencing this. Even following the most basic example here: [https://nextjs.org/docs/basic-features/script](https://nextjs.org/docs/basic-features/script) with a custom _document.jsx `beforeInteractive` scripts never are appended to the head and never execute.

**Comment by negprof** [comment]
I do have the same issue on next.js 13

**Comment by sanghunka** [comment]
Same issue.

I solved this issue by using onLoad rather than dangerouslySetInnerHTML

```js
<Script
  id="gtm"
  strategy="afterInteractive"
  // dangerouslySetInnerHTML={{ __html: `console.log('hi')`}}  //doesn't work
  onLoad={() => {
    console.log('hi') //It works
  }}
/>
```

**Reply by tappress** [reply]
You can't do this with server components though, since events (onLoad) are not available there.

**Comment by OZZlE** [comment]
<Script
              src="/globalassets/adobe/30e54a9d6c99/28871c755140/launch-32a177491a9c.js"
              async
              strategy="afterInteractive"
              onLoad={() => {
                debugger;
              }}
            />

Does nothing, amazing this tag so well documented and works 0% of the time? :D

**Reply by icyJoseph** [reply]
In what context are you using it? Funny how it works 100% of the time for some, and it totally doesn't for others... Is this in a Server Component?

**Reply by OZZlE** [reply]
<LocalizationProvider dateAdapter={AdapterDayjs} adapterLocale={props.locale}>
        <Head>
          <title>{title}</title>
          <meta name="robots" content="noindex, nofollow" />
          <meta name="viewport" content="width=device-width, initial-scale=1" />
          <Script src="/globalassets/adobe/30e54a9d6c99/28871c755140/launch-32a177491a9c.js" />
        </Head>

It's Adobes launch script for tracking, perhaps it collides with something this Script component tries to do? It doesn't log to console at least like it should

**Comment by jesslark** [comment]
Similar issue in the tutorial, on this page:
https://nextjs.org/learn/basics/assets-metadata-css/third-party-javascript

I can't get the Script component to work. There is no error; nothing happens at all. I am following the tutorial on a mac, using Chrome to view the output.

To troubleshoot I:
- checked the network tab and don't see the SDK
- waited a "while" in case it is VERY lazy, still nothing in the network tab, and no console
- reverted to the previous code to check to see if it loaded with that, and it did (the previous code was a regular lower case script tag with the same SDK link)
- added a console separate from the Script tag, it did console correctly
- moved the Script component to the index file, still nothing. No error, just, nothing.

Code:

```
import Head from "next/head"
import Link from "next/link"
import Script from "next/script"

export default function FirstPost() {
  return (
    <>
      <Head>
        <title>First POAST</title>
        <Script
          src="https://connect.facebook.net/en_US/sdk.js"
          strategy="lazyOnload"
          onLoad={() =>
            console.log(`script loaded correctly, window.FB has been populated`)
          }
        />
        {console.log("WHOA IS THIS EVEN WORKING")}
      </Head>
      <h1>First Post</h1>
      <h2>
        <Link href="/">Back to home</Link>
      </h2>
    </>
  )
}


```

Dependencies include:

Node v: 20.4.0

"dependencies": {
"next": "latest",
"react": "18.2.0",
"react-dom": "18.2.0"
}

**Comment by maxtheaxe** [comment]
Also not working for me. Strategy change to `lazyOnload` does not fix it, though usage of a native `<script>` tag does. It does seem to work on `_document.tsx` (at least with `dangerouslySetInnerHTML`, but in this case, I only want to load specific scripts on specific pages, which I do using `<Head>` in page definitions. Happy to do more testing if that would help resolve this issue.

**Comment by bezreyhan** [comment]
After running into this issue I noticed that [the docs](https://nextjs.org/docs/pages/building-your-application/optimizing/scripts) say that the `Script` component has to be used inside of `_app`: 

> To load a third-party script for all routes, import next/script and include the script directly in your custom _app

**Comment by maxtheaxe** [comment]
> After running into this issue I noticed that [the docs](https://nextjs.org/docs/pages/building-your-application/optimizing/scripts) say that the `Script` component has to be used inside of `_app`:
> 
> > To load a third-party script for all routes, import next/script and include the script directly in your custom _app

Yes, but that's only if you want to load the script for all routes. I'm not seeing any indication that that's its exclusive functionality.

**Comment by jdgarcia** [comment]
Not sure if this is related but I have a `<Script strategy='beforeInteractive' src='/js/context.js' />` in `app/layout.ts` and the request for it gets cancelled on some (but not all) pages.

**Comment by bilalbutt044** [comment]
i am having same issue in nextjs version `12.1.6 `.  Nextjs's  <Script /> tag is not loading in _document.js file. i have to use native <script/> tag or put <Script /> component in _app.js file.   

NOTE: Nexjts recommend not to use <Script /> in _document.js file as _document.js file runs on server and <Script /> tag contains Frontend functionality
https://nextjs.org/docs/messages/no-script-in-document

**Comment by Armadillidiid** [comment]
adding `beforeInteractive` as the strategy worked for me.

**Comment by theisof** [comment]
Started to work for me when I moved `<Script>` out of `<Head>` 🤷‍♂️ (in _app.tsx)

**Reply by 17Amir17** [reply]
worked when I moved to a regular head 🤷‍♂️

**Comment by medjabir** [comment]
@theisof Same here :(

**Comment by mk48** [comment]
I don't know whether this will be helpful
I had the same issue, but my Ad-blocker blocked the tracking script (umami).
I used NextJs, tried to add Umami analytics using Script feature from NextJs.

![image](https://github.com/user-attachments/assets/d2550def-d3c4-4d5f-8bbb-6b4f226a7c3c)
